audit-checklists/Arbitrum.md

```diff
@@ -0,0 +1,95 @@
+
+> Look for vulnerabilities in protocol that can come out of retryable tickets executing in different order than they were submitted
+
+Bug examples: [1](https://docs.arbitrum.io/assets/files/2024_08_01_trail_of_bits_security_audit_custom_fee_token-7ce514634632f4735a710c81b55f2d27.pdf#page=12)
+
+### Incorrect cross-chain authentication
+
+Arbitrum, as well as other L2s, applies [address aliasing](https://docs.arbitrum.io/how-arbitrum-works/l1-to-l2-messaging#address-aliasing) to the retryable ticket's sender's address when sender is a smart contract. Sometimes L2 contract has a function which is only callable by L1 counterpart contract. In such cases it is important to apply alias when checking the sender. If sender is not aliased before check, function won't be callable at all. Check should look something like this:
+
+```solidity
+require(msg.sender == AddressAliasHelper.applyL1ToL2Alias(l1Counterpart), "only counterpart");
+```
+
+> When retryable ticket's sender address is checked make sure alias is applied where needed
+
+Bug examples: [1](https://solodit.cyfrin.io/issues/msgsender-has-to-be-un-aliased-in-l2blastbridgefinalizebridgeethdirect-spearbit-none-draft-pdf)
+
+### Spending the nonce
+
+There are different ways in which retryable ticket can fail on L2. If the L2 gas price provided is too low, ie. due to gas price spike, retryable ticket will not be scheduled for auto-redemption. On the other hand, auto-redemption can be scheduled, but fail due to too low gas limit provided (TX runs out of gas). In the latter case (aliased) sender's nonce will be spent. This can have important implication if for example L1 contract is used to deploy contract on L2 and predict the address of the new contract. If gas limit provided for deployment is too low, nonce 0 will be burned. Deployment can still be done by doing the manual redemption of retryable ticket, however address of the deployed contract will be different than the one initially predicted by L1 contract because nonce 1 was used for deployment instead of nonce 0.
+
+> When L1 contract is used to deploy contracts to L2 look for any hidden assumptions about the deployment nonce
+
+### Permission to cancel the retryable ticket
+
+When creating new retryable ticket one of the parameters that is provided to the `Inbox::createRetryableTicket` is `callValueRefundAddress`. This is the address where `l2CallValue` is refunded to in the case that retryable ticket execution fails on L2. But additionally `callValueRefundAddress` has the permission to cancel the ticket, making it permanently unclaimable. It is important to keep this in mind in case protocol is designed in a way where this can be exploited. Here's an example of such scenario:
```

Comment:
Same comment as above. Try to put links to the contract functions.

Reply:
Added